x/oob/writer.go

```diff
@@ -0,0 +1,118 @@
+package oob
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/x/sockopt"
+)
+
+type oobWriter struct {
+	conn        *net.TCPConn
+	opts        sockopt.TCPOptions
+	resetTTL    sync.Once
+	setTTL      sync.Once
+	oobPosition int64
+	sd          SocketDescriptor
+	oobByte     byte // Byte to send as OOB
+	disOOB      bool // Flag to enable disOOB mode
+	delay       time.Duration
+}
+
+var _ io.Writer = (*oobWriter)(nil)
+
+type oobWriterReaderFrom struct {
+	*oobWriter
+	rf io.ReaderFrom
+}
+
+// NewWriter creates an [io.Writer] that sends an OOB byte at the specified "oobPosition".
+// If disOOB is enabled, it will apply the --disOOB strategy.
+// "oobByte" specifies the value of the byte to send out-of-band.
+func NewWriter(
+	conn *net.TCPConn,
+	opts sockopt.TCPOptions,
+	oobPosition int64,
+	oobByte byte,
+	disOOB bool,
+	delay time.Duration,
+) io.Writer {
+	return &oobWriter{conn: conn, opts: opts, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB, delay: delay}
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if w.oobPosition > 0 && w.oobPosition < int64(len(data))-1 {
+		firstPart := data[:w.oobPosition+1]
+		secondPart := data[w.oobPosition:]
+
+		// Split the data into two parts
+		tmp := secondPart[0]
+		secondPart[0] = w.oobByte
+
+		var oldTTL int
+		if w.disOOB {
+			w.setTTL.Do(func() {
+				oldTTL, err = w.opts.HopLimit()
```

Comment:
Take a look at the suggestion in https://github.com/Jigsaw-Code/outline-sdk/pull/329/files#r1840884687 that could simplify this and reduce code duplication.